File: src/main/java/commands/listeners/OnSelectionMenuListener.java
package commands.listeners;

import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import commands.CommandListenerMeta;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.events.interaction.SelectionMenuEvent;

public interface OnSelectionMenuListener extends OnInteractionListener {

    boolean onSelectionMenu(SelectionMenuEvent event) throws Throwable;

    default CompletableFuture<Long> registerSelectionMenuListener(Member member) {
        return registerSelectionMenuListener(member, true);
    }

    default CompletableFuture<Long> registerSelectionMenuListener(Member member, boolean draw) {
        return registerInteractionListener(member, OnSelectionMenuListener.class, draw, this::onSelectionMenuOverridden);
    }

    default CompletableFuture<Long> registerSelectionMenuListener(Member member, Function<SelectionMenuEvent, CommandListenerMeta.CheckResponse> validityChecker, boolean draw) {
        return registerInteractionListener(member, OnSelectionMenuListener.class, draw, this::onSelectionMenuOverridden, validityChecker);
    }

    default void processSelectionMenu(SelectionMenuEvent event) {
        processInteraction(event, this::onSelectionMenu);
    }

    default void onSelectionMenuOverridden() throws Throwable {
    }

}

File: src/main/java/commands/listeners/OnSelectionMenuListener.kt
package commands.listeners

import commands.CommandListenerMeta.CheckResponse
import mysql.hibernate.EntityManagerWrapper
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent
import java.util.concurrent.CompletableFuture

interface OnSelectMenuListener : OnInteractionListener {

    @Throws(Throwable::class)
    fun onSelectMenu(event: StringSelectInteractionEvent): Boolean

    fun registerSelectMenuListener(member: Member): CompletableFuture<Long> {
        return registerSelectMenuListener(member, true)
    }

    fun registerSelectMenuListener(member: Member, draw: Boolean): CompletableFuture<Long> {
        return registerInteractionListener(member, OnSelectMenuListener::class.java, draw) { onSelectMenuOverridden() }
    }

    fun registerSelectMenuListener(member: Member, validityChecker: (StringSelectInteractionEvent) -> CheckResponse, draw: Boolean): CompletableFuture<Long> {
        return registerInteractionListener(member, OnSelectMenuListener::class.java, draw, { onSelectMenuOverridden() }, validityChecker)
    }

    fun processSelectMenu(event: StringSelectInteractionEvent, entityManager: EntityManagerWrapper) {
        processInteraction(event, entityManager) { onSelectMenu(it) }
    }

    @Throws(Throwable::class)
    fun onSelectMenuOverridden() {
    }

}